Implemented the new mailbox action menu

divyamrastogi opened the pull request:
This feature will enable rendering the mailbox actions as an ellipsis action menu.

Based on the type of Mailbox and available actions, either Gmail's or Titan's action menu is being shown to the user.

#### Changes proposed in this Pull Request

* Added a Mailbox action menu consisting of email-related actions for Titan and Gmail.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be beneficial when the change is visual.
-->

1. Open the branch on your local server and go to `Upgrades` -> `Domains.`
2. Click any email account to open its Domain view
3. Click on "Manage your email accounts."
4. Click on the `EllipsisMenu` on the right of your mailbox to see.

**Before:**
![image](https://user-images.githubusercontent.com/1598017/118120747-fdc43180-b40d-11eb-98cd-4c933710c618.png)

**After (Gmail Action Menu)**
![image](https://user-images.githubusercontent.com/1598017/118120814-17657900-b40e-11eb-8261-bc875c60c618.png)

**After (Titan Action Menu)**
![image](https://user-images.githubusercontent.com/1598017/118120879-35cb7480-b40e-11eb-9e84-ca85eda0a3f0.png)

cc: @daledupreez

divyamrastogi committed:
Implemented the new mailbox action menu

We will conditionally render Titan's action menu and Gmail's Action menu based on type of mail box.

divyamrastogi committed:
Implemented the new mailbox action menu

We will conditionally render Titan's action menu and Gmail's Action menu based on type of mail box.

divyamrastogi committed:
Implemented the new mailbox action menu

We will conditionally render Titan's action menu and Gmail's Action menu based on type of mail box.

divyamrastogi committed:
Implemented the new mailbox action menu

We will conditionally render Titan's action menu and Gmail's Action menu based on type of mail box.

divyamrastogi committed:
Hide external links icon in mailbox action-menu

divyamrastogi committed:
Hide external links icon in mailbox action-menu

divyamrastogi committed:
Hide external links icon in mailbox action-menu

divyamrastogi committed:
Hide external links icon in mailbox action-menu

divyamrastogi committed:
Add custom popover class to mailbox action menu

divyamrastogi committed:
Add custom popover class to mailbox action menu

divyamrastogi committed:
Add custom popover class to mailbox action menu

divyamrastogi committed:
Add custom popover class to mailbox action menu

divyamrastogi committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into add/mailbox-action-menu

divyamrastogi committed:
No need to pass down domain to TitanActionMenu

divyamrastogi committed:
No need to pass down domain to TitanActionMenu

divyamrastogi committed:
No need to pass down domain to TitanActionMenu

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -3,6 +3,8 @@
 import React from 'react';
 import classNames from 'classnames';
+import EllipsisMenu from 'calypso/components/ellipsis-menu';
+import PopoverMenuItem from 'calypso/components/popover/menu-item';
```
Minor note: these are internal dependencies, so should be added in the second block of dependencies below.

divyamrastogi replied:
Done. I need to make this a habit!

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -20,9 +22,29 @@ import {
+import titanCalendarIcon from 'calypso/assets/images/email-providers/titan/services/calendar.svg';
+import titanContactsIcon from 'calypso/assets/images/email-providers/titan/services/contacts.svg';
+import titanMailIcon from 'calypso/assets/images/email-providers/titan/services/mail.svg';
+import Gridicon from 'calypso/components/gridicon';
```
Nit: to keep alphabetical order, this should follow the `googleSlidesIcon` import.

divyamrastogi replied:
Missed this. Done now.

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -109,7 +131,122 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+				href={ getGmailUrl( domainName ) }
+			>
+				<img src={ gmailIcon } alt={ translate( 'Gmail icon' ) } />
+				View Gmail
```
All the text values for the Google menu need to be translated.

divyamrastogi replied:
Done. Used translate on the string.

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -109,7 +131,122 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+		</EllipsisMenu>
+	);
+};
+const GmailActionMenu = ( { domainName } ) => {
```
Nit: please add a space between this const and the previous code.

divyamrastogi replied:
Done!

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -139,24 +276,31 @@ function EmailPlanMailboxesList( { mailboxes, isLoadingEmails } ) {
-							comment: 'Email user role displayed as a badge',
-						} ) }
-					</Badge>
+				{ hasTitanMailWithUs( domain ) ? (
+					<TitanActionMenu />
+				) : (
+					<GmailActionMenu domainName={ domain.name } />
 				) }
```
_Warning: I had a lot of thoughts that got dumped in this comment!_

This component is also used to show email forwards, so I think its going to be simpler to push this logic into a separate function, either a generic/local component (i.e. `const ActionMenu = ( { domain } ) => {}`), or as a helper function (i.e. `const getActionMenu = ( domain ) => {}`), and then drop that into this main JSX block.

I am also thinking that it may be simpler to have a single location where we add an `EllipsisMenu` -- we can then use helper functions to add the relevant menu items, and have only one "menu" component.

I am thinking the code could look something like this:
```js
const TitanMenuItems = () => {
	const translate = useTranslate();

	return (
		<>
			<PopoverMenuItem
				href="https://wp.titan.email/mail/"
				className="email-plan-mailboxes-list__mailbox-action-menu-item"
				isExternalLink={ true }
			>
				<img src={ titanMailIcon } alt={ translate( 'Titan Mail icon' ) } />
				{ translate( 'Mail', {
					comment: 'This refers to the Email application (i.e. the webmail) of Titan',
				} ) }
			</PopoverMenuItem>
			<PopoverMenuItem
				href="https://wp.titan.email/calendar/"
				className="email-plan-mailboxes-list__mailbox-action-menu-item"
				isExternalLink={ true }
			>
				<img src={ titanCalendarIcon } alt={ translate( 'Titan Calendar icon' ) } />
				{ translate( 'Calendar', {
					comment: 'This refers to the Calendar application of Titan',
				} ) }
			</PopoverMenuItem>
			<PopoverMenuItem
				href="https://wp.titan.email/contacts/"
				className="email-plan-mailboxes-list__mailbox-action-menu-item"
				isExternalLink={ true }
			>
				<img src={ titanContactsIcon } alt={ translate( 'Titan Contacts icon' ) } />
				{ translate( 'Contacts', {
					comment: 'This refers to the Contacts application of Titan',
				} ) }
			</PopoverMenuItem>
		</>
	);
};

const GoogleMenuItems = ( { domain } ) => {
   // ...
};

const getMenuItems = ( domain ) => {
	if ( hasTitanMailWithUs( domain ) ) {
		return <TitanMenuItems />;
	}

	if ( hasGSuiteWithUs( domain ) ) {
		return <GoogleMenuItems domain={ domain } />;
	}

	// TODO: Add email forwarding options or remove them
	return null;
};

const ActionMenu = ( { domain } ) => {
	const actionMenuItems = getMenuItems( domain );

	if ( ! actionMenuItems ) {
		return null;
	}

	return (
		<EllipsisMenu
			className="email-plan-mailboxes-list__mailbox-action-menu"
			popoverClassName="email-plan-mailboxes-list__mailbox-action-menu-popover"
		>
			{ actionMenuItems }
		</EllipsisMenu>
	);
};
```

It may not be for this PR, but all the `PopoverMenuItem` boilerplate for the menu items also bothers me. It may be worth defining the data here via arrays and then mapping that to components. e.g.
```js
const getTitanMenuItems = ( translate ) => {
	return [
		{	
			href: "https://wp.titan.email/mail/",
			image: titanMailIcon,
			imageAltText: translate( 'Titan Mail icon' ),
			title: translate( 'Mail', {
				comment: 'This refers to the Email application (i.e. the webmail) of Titan',
			} ),
		},
		{
			href: "https://wp.titan.email/calendar/",
			image: titanCalendarIcon,
			imageAltText: translate( 'Titan Calendar icon' ),
			title: translate( 'Calendar', {
				comment: 'This refers to the Calendar application of Titan',
			} ),
		},
		{
			href: "https://wp.titan.email/contacts/",
			image: titanContactsIcon,
			imageAltText: translate( 'Titan Contacts icon' ),
			title: translate( 'Contacts', {
				comment: 'This refers to the Contacts application of Titan',
			} ),
		}
	];
};

const getMenuItems = ( domain, translate ) => {
	if ( hasTitanMailWithUs( domain ) ) {
		return getTitanMenuItems( translate );
	}
	// ...
};

const ActionMenu = ( { domain } ) {
	const translate = useTranslate();
	const menuItems = getMenuItems( domain, translate );
	if ( ! menuItems ) {
		return null;
	}
	return (
		<EllipsisMenu>
			{ menuItems.map( function( menuItem ) {
				const { image, imageAltText, href,  title } = menuItem;

				// We likely need a key here -- might be best to use the `href`
				return (
					<PopoverMenuItem
						className="email-plan-mailboxes-list__mailbox-action-menu-item"
						isExternalLink
						href={ href }
					>
						<img src={ image } alt={ imageAltText } />
						{ title }
					</PopoverMenuItem>
				);
			} ) }
		</EllipsisMenu>
	);
};
```

divyamrastogi replied:
Created a mappable array out of the menu items as you've suggested.

I needed to understand more about "showing email forwards". Is it something I missed from the designs?

daledupreez replied:
I think so -- you can take a look at #52750, especially [this comment](https://github.com/Automattic/wp-calypso/pull/52750#issuecomment-840593024) for a screenshot and some discussion.

I am wondering whether these designs were only in the Figma and not in the post.

daledupreez reviewed (commented):
Overall, this is looking good, but I do have some comments for you.

I also know that this is likely to hit some merge issues as a result of #52750, which I merged earlier today. LMK if you need any help with the rebase.

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -109,7 +131,122 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+			<PopoverMenuItem
+				className="email-plan-mailboxes-list__mailbox-action-menu-item"
+				isExternalLink={ true }
+				href={ getGmailUrl( domainName ) }
```
Another thought: I think we should add some of these links (like Gmail) using the full email address, not just the domain name.
I am not sure if Titan also supports that.

divyamrastogi replied:
By including the full email address, do you mean we need to pass the `email` instead of `domainName` to the `getGmailUrl` function? If you're talking about email address auto-fill, that seemed to be happening irrespective of whether we pass down `email` or `domainName` in the `getGmailUrl` method.

Pretty sure Titan doesn't support autofill through the generated URL as of now, but I can ask someone to do that for us.

daledupreez replied:
Not urgent if Titan doesn't support that yet.

But I think the complication will be when there are multiple email addresses -- specifying the email address is likely to be the right option if we're showing this menu alongside each mailbox name.

daledupreez replied:
And yes, I did mean passing `email` to the `getGmailUrl()` function for _all_ the Google links.

divyamrastogi replied:
Done, now passing `email` instead of `domainName` to the GSuite utilities.

divyamrastogi committed:
Organize component imports in alphabetical order

divyamrastogi committed:
Organize component imports in alphabetical order

divyamrastogi committed:
Segregate logic from ActionMenu component

Reuse the same EllipsisMenu and PopoverMenuItems components and create helper functions to decide whether Titan of Gmail menu items are shown.

divyamrastogi committed:
Segregate logic from ActionMenu component

Reuse the same EllipsisMenu and PopoverMenuItems components and create helper functions to decide whether Titan of Gmail menu items are shown.

divyamrastogi committed:
Merge branch 'add/mailbox-action-menu' of github.com:talk-to/wp-calypso into add/mailbox-action-menu

divyamrastogi committed:
Merge branch 'add/mailbox-action-menu' of github.com:talk-to/wp-calypso into add/mailbox-action-menu

divyamrastogi committed:
Remove external link

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -11,6 +11,8 @@ import { useTranslate } from 'i18n-calypso';
  * Internal dependencies
  */
+import EllipsisMenu from 'calypso/components/ellipsis-menu';
+import PopoverMenuItem from 'calypso/components/popover/menu-item';
```
Nit: Please push this lower down the list so it's in alphabetical order.

jsnajdr commented on client/my-sites/email/email-management/style.scss:
```diff
@@ -149,6 +149,22 @@
 	margin-top: 15px;
 }
+.email-plan-mailboxes-list__mailbox-action-menu-popover {
+	.email-plan-mailboxes-list__mailbox-action-menu-item {
```
Nesting these selectors is probably not necessary. The long BEM-style names should make them unique enough.

divyamrastogi replied:
More specificity was needed to override some of the inbuilt styles in MenuPopover components. I needed to hide the external link icon while preserving its behaviour as an external link to adhere to the given designs.

jsnajdr replied:
In that case I would do a different selector, like:
```css
.native__menu-item.my-specific-ui__menu-item
```
That has the same specificity and makes it more clear that we are overriding a style and which one it is.

divyamrastogi replied:
Right, makes more sense!

jsnajdr commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -136,7 +159,125 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+			className="email-plan-mailboxes-list__mailbox-action-menu"
+		>
+			{ menuItems.map( function ( menuItem ) {
+				const { image, imageAltText, href, title } = menuItem;
```
style nit: destructuring `menuItem` in the argument list would make the callback more concise.

divyamrastogi replied:
Done

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -136,7 +159,125 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+		return getGmailMenuItems( domain.name, translate );
+	}
+
+	return null;
```
Thanks for splitting this out! I think the only item we currently have for this menu is an "Edit" link that should call the `emailManagementForwarding()` helper function to get the path to manage forwarding for the domain.

I am thinking it may be worth pushing that to a follow-up PR so we can get this one shipped, as that function requires some additional data derived from the state.

divyamrastogi replied:
All right, will need some context here though.

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -136,7 +159,125 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+	return null;
+};
+
+const ActionMenu = ( { domain } ) => {
```
I think we should be passing in the `mailbox` here as well. The Google links only include the domain, but not the email, despite the user having clicked on a particular email address to launch the option.

This diverges from the previous logic, as those links were displayed at the domain level, not at the mailbox level.

daledupreez commented on client/my-sites/email/email-management/style.scss:
```diff
@@ -164,6 +180,14 @@
 		border-left-color: var( --color-error-50 );
 	}
+	.email-plan-mailboxes-list__mailbox-list-item-email {
+		display: flex;
+		align-content: space-between;
+		flex-basis: 100%;
```
These styles seem to break the mobile layout for the email forwarding screen:

<img width="667" alt="Screenshot 2021-05-17 at 09 20 42" src="https://user-images.githubusercontent.com/3376401/118447631-4ff8a180-b6f1-11eb-9ab4-97e87b38a420.png">

The related styles were added in #52750. I am thinking it may be simpler to use `position: absolute;` with `top: 0;` and `right: 24px;` to get this menu into the right location regardless of how the other items in the card shift around. An example of this approach is in the icons shown to the right of the "Add new mailboxes" and "Manage all mailboxes" menu items at the bottom of the email plan page.

divyamrastogi replied:
Done, used `top:50%; translateY(-50%);` along with the above suggestions to make it foolproof.

daledupreez replied:
OK -- I think I tracked the issue down -- with these changes to the menu, we no longer need the `<div className="email-plan-mailboxes-list__mailbox-list-item-email">` wrapper div and its associated styles. I _think_ you added these to help with placing the menu, but that's no longer needed AFAICT. (I tested this locally by removing the div and undoing the relevant style changes.)

daledupreez reviewed (commented):
From my side, this is looking really good.

I think we should defer adding the email forward menu items to a later PR, but I think we do have two things to address before we can ship this:
1. Change the CSS to right-align the ellipsis menu
2. Update the Google links to use the mailbox

divyamrastogi committed:
Import components in alphabetical order

stephanethomas commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -136,7 +159,125 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+	}
+
+	if ( hasGSuiteWithUs( domain ) ) {
+		return getGmailMenuItems( domain.name, translate );
```
Could we please rename `getGmailMenuItems` to `getGSuiteMenuItems`? This would be more accurate as this method lists not only Gmail but also some other apps that are part of G Suite.

stephanethomas commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -136,7 +159,125 @@ const getActionsForMailbox = ( mailbox, translate, dispatch ) => {
+			image: googleAdminIcon,
+			imageAltText: translate( 'Google Admin icon' ),
+			title: translate( 'View Admin' ),
+		},
```
We should not display that particular link for non-admin users, as they don't have access to the Google Admin Console. This can be addressed in another pull request though.

divyamrastogi committed:
Destructure menuItem within map function

divyamrastogi committed:
Rename getGmailMenuItems to getGSuiteMenuItems

divyamrastogi committed:
Remove custom class from PopoverMenu

divyamrastogi committed:
Make syntax more precise

divyamrastogi committed:
Correcting style for external link

divyamrastogi committed:
Pass down email to GSuite utilities

divyamrastogi committed:
Pass down email to Titan utils for email autofill

divyamrastogi committed:
Use position absolute for the action menu

divyamrastogi committed:
Rename domainName to email as we pass only email

daledupreez commented on client/lib/titan/index.js:
```diff
@@ -6,4 +6,5 @@ export { getTitanMailboxRenewalCost } from './get-titan-mailbox-renewal-cost';
 export { getTitanMailOrderId } from './get-titan-mail-order-id';
 export { getTitanProductName } from './get-titan-product-name';
 export { getTitanSubscriptionId } from './get-titan-subscription-id';
+export { getTitanEmailUrl, getTitanCalendarlUrl, getTitanContactsUrl } from './get-titan-urls';
```
Non-blocking nit: please export these in alphabetical order.

daledupreez commented on client/lib/titan/get-titan-urls.js:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Available Titan Apps
+ */
+export const TITAN_APPS = {
```
I don't think `TITAN_APPS` should be exported -- callers should only care about the more specific exported functions.

daledupreez commented on client/my-sites/email/email-management/home/email-plan-mailboxes-list.jsx:
```diff
@@ -41,7 +41,12 @@ import googleDriveIcon from 'calypso/assets/images/email-providers/google-worksp
 import Gridicon from 'calypso/components/gridicon';
-import { hasTitanMailWithUs } from 'calypso/lib/titan';
+import {
+	getTitanEmailUrl,
```
Nit: Please shift this earlier to keep the alphabetical import order.

daledupreez commented on client/my-sites/email/email-management/style.scss:
```diff
@@ -150,6 +150,13 @@
+.email-plan-mailboxes-list__mailbox-action-menu {
+	position: absolute;
+    right: 24px;
+    top: 50%;
```
The alignment/whitespace here seems a bit off.

divyamrastogi replied:
Replaced tab with spaces here.

daledupreez reviewed (commented):
Other than some minor nits, this is looking good!

Unfortunately, we still need to remove the additional wrapper div in the card -- it's causing issues for mobile layouts.

divyamrastogi committed:
Organize titan utils in ascending order

divyamrastogi committed:
Remove item-email wrapper

divyamrastogi committed:
Remove export from TITAN_APPS

daledupreez commented on client/lib/titan/get-titan-urls.js:
```diff
@@ -25,14 +25,14 @@ function getTitanUrl( email, app = TITAN_APPS.EMAIL ) {
 
-export function getTitanEmailUrl( email ) {
+export function getTitanCalendarlUrl( email ) {
 	return getTitanUrl( email, TITAN_APPS.EMAIL );
```
Shouldn't this be `TITAN_APPS.CALENDAR`? (Same applies below for the contacts URL.)

daledupreez reviewed (commented):
LGTM! The only issue at the moment for this PR is fixing the constants for the Titan link construction.

daledupreez reviewed (approved):
My approval got switched to a comment by mistake.

divyamrastogi committed:
Use the correct links in order

daledupreez reviewed (approved):
Approving after the final commit -- not sure if that's blocking some of the tests from running.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5919861

Thank you @divyamrastogi for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5919861).
